src/main/ruby/core/truffle/cext.rb

```diff
@@ -1924,6 +1924,189 @@ def rb_sprintf(format, *args)
     f = format.gsub(/%(?:\+)?l.\v/, '%s')
     sprintf(f, *args) rescue raise ArgumentError, "Bad format string #{f}."
   end
+
+  class LocalRefs
+
+    attr_reader :parent
+    attr_reader :refs
+    attr_reader :id
+
+    def initialize(parent)
+      @parent = parent
+      @refs = []
+      if parent
+        @id = parent.id + 1
+      else
+        @id = 0
+      end
+    end
+
+    def add_ref(obj)
+      res = calc_ref_id
+      @refs << obj
+      res
+    end
+
+    def marker
+      0xbaddecaf000000
+    end
+    
+    def get_ref(ref)
+      id_part = (ref >> 16) & 0xff
+      index_part = ref & 0xffff
+      frame = self
+      while (frame && frame.id != id_part)
+        frame = frame.parent
+      end
+      if frame && index_part < frame.refs.size
+        frame.refs[index_part]
+      else
+        p "Bad get ref #{ref.to_s(16)}."
+        puts caller[0..8]
+        nil
+      end
+    end
+
+    def calc_ref_id
+      index_part = @refs.size
+      id_part = (@id & 0xff) << 16
+      index_part | id_part | marker
+    end
+  end
+
+  class GlobalRefs
+    attr_reader :refs
+
+    def initialize
+      @refs = {}
+      @count = 1 # Avoid returning something that might be interpreted as null.
+      @deletes = 0
+    end
+
+    def marker
+      0xdeadbeef0000000
+    end
+    
+    def add_ref(obj)
+      ref = @count
+      @refs[ref] = obj
+      @count += 1
+        
+      ref | marker
+    end
+
+    def get_ref(ref)
+      ref = ref & 0xfffffff
+      unless @refs.include?(ref)
+        return nil
+      end
+      @refs[ref]
+    end
+
+    def release_ref(ref)
+      @refs.delete(ref)
+      @deletes += 1
+    end
+  end
+  
+  class WeakRefs
+    attr_reader :refs
+
+    def initialize
+      @refs = {}
+      @count = 1 # Avoid returning something that might be interpreted as null.
+      @deletes = 0
+    end
+
+    def marker
+      0xcafefade0000000
+    end
+    
+    def add_ref(obj)
+      ref = @count
+      @refs[ref] = WeakRef.new(obj)
+      @count += 1
+        
+      ref | marker
+    end
+    
+    def get_ref(ref)
+      ref = ref & 0xfffffff
+      unless @refs.include?(ref)
+        return nil
+      end
+      res = @refs[ref]
+      unless res.weakref_alive?
+        return nil
+      end
+      res.__getobj__
```

Comment:
Is it possible that `weakref_alive?` returns true but then there is a GC before `__getobj__`?